transaction.go

```diff
@@ -0,0 +1,489 @@
+package factom
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/Factom-Asset-Tokens/factom/varintf"
+)
+
+type FactoidTransaction struct {
+	// TODO: The header is usually at the top level, is this ok?
+	FactoidTransactionHeader
+
+	FCTInputs  []FactoidTransactionIO        `json:"inputs"`
+	FCTOutputs []FactoidTransactionIO        `json:"fctoutputs"`
+	ECOutputs  []FactoidTransactionIO        `json:"ecoutputs"`
+	Signatures []FactoidTransactionSignature `json:"signatures"`
+}
+
+type FactoidTransactionHeader struct {
+	// TransactionID is not in the marshalled binary
+	TransactionID *Bytes32 `json:"txid"`
+
+	Version uint64 `json:"version"`
+	// Timestamp is accurate to the millisecond
+	Timestamp time.Time `json:"timestamp"`
+
+	InputCount     uint8 `json:"inputcount"`
+	FCTOutputCount uint8 `json:"fctoutcount"`
+	ECOutputCount  uint8 `json:"ecoutcount"`
+}
+
+type FactoidTransactionIOs []FactoidTransactionIO
+
+type FactoidTransactionIO struct {
+	Amount uint64 `json:"amount"`
+	// Address can be an SHA256d(RCD) for FCT in/out, or a public key for EC out.
+	// It is the encoded bytes into the human readable addresses
+	Address *Bytes32 `json:"address"`
+
+	// TODO: Do you want to also put the human readable address when we unmarshal?
+}
+
+type FactoidTransactionSignature struct {
+	// SHA256d(RCD) == FactoidIOAddress for the inputs
+	ReedeemCondition RCD   `json:"rcd"`
+	SignatureBlock   Bytes `json:"amount"`
+}
+
+// IsPopulated returns true if f has already been successfully populated by a
+// call to Get. IsPopulated returns false if f.FCTInputs, or f.Signatures are
+// nil, or if f.Timestamp is zero.
+func (f FactoidTransaction) IsPopulated() bool {
+	return f.FCTInputs != nil && // Although a coinbase has 0 inputs, this array should not be nil
+		f.Signatures != nil &&
+		f.Timestamp != time.Time{}
```

Comment:
time.Time has an IsZero function for this purpose

Reply:
TIL